Design/spectrum_number_and_workspace_index_abstraction.md

```diff
@@ -0,0 +1,72 @@
+# Spectrum Number and Workspace Index Abstraction Layer
+
+
+## Motivation
+
+There are two main motivations for this design proposal:
+
+1. We need to provide a consistent user interface that allows users to specify which spectrums in a workspace an algorithm should modify.
+2. Developers of algorithms should not have to deal with conversions between spectrum number and other index types.
+
+
+## Current situation and future problems
+
+#### Definitions:
+1. Workspace index = linear index starting at zero, corresponds to memory offset in vector of spectra
+2. Spectrum number = user defined index for spectra in certain workspace
+
+#### Current situation:
+
+- Conversion methods between index types (spectrum number and workspace index) are part of the workspace interface.
+- There is no consistent way of defining index ranges or lists for algorithms. This is done individually in each algorithm, and properties may be defined for spectrum number or workspace indices, and sometimes the interpretation of index properties even seems to depend on the facility.
+- Validation of index ranges and behavior on errors such as out-of-range indices is likewise done individually in all algorithms. Algorithms behave inconsistently depending on errors (throw and error, print a warning, ignore silently). There are bugs in some range validations.
+
```

Comment:
I might also highlight the problem of the hard-coded assumption about mapping. There are a number of areas in the codebase where the spectrum numbers are assumed to start from 1. This is usually the case at ISIS, but if it was different in the file (and it could be) it would be a nasty problem to track down.